Shell formatting and linting

bitdivine opened the pull request:
# Motivation
A shell script just failed because it was written in bash, had no shebang, and was executed in dash.

So I'm adding shellcheck and also shfmt, while I am about it.

Note: This PR is against RC1.  I could apply it to main as well, possibly with minor changes if the set of shell scripts requiring changes has diverged.

# Changes
- Add shellcheck and shfmt, both as github checks and as scripts to be executed locally.
  - Note:  I have not made github automatically commit shfmt output, as reformatting shell is riskier than reformatting most things.
  - Note: I have made the indent 2 spaces rather than the 4 used in the main ic repo shfmt, to match the existing code.
  - Note: I have kept the formatting scripts simple and lightweight rather than using the more sophisticated approach used in the main ic repo.

nmattia commented on frontend/ts/src/proto/update_proto.sh:
```diff
@@ -1,22 +1,23 @@
+#!/usr/bin/env bash
```
```suggestion
#!/usr/bin/env bash
set -euo pipefail
```

bitdivine replied:
I agree but did not want to impose too much.  If people are fine with it I'll happily add it.

nmattia commented on frontend/svelte/nns-js/proto/update_proto.sh:
```diff
@@ -1,14 +1,15 @@
+#!/usr/bin/env bash
```
```suggestion
#!/usr/bin/env bash
set -euo pipefail
```

nmattia commented on deploy.sh:
```diff
@@ -1,4 +1,5 @@
+#!/usr/bin/env bash
```
```suggestion
#!/usr/bin/env bash
set -euo pipefail
```

nmattia commented on bin/lint-sh:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
```
```suggestion
#!/usr/bin/env bash
set -euo pipefail
```

nmattia replied:
... and worth adding to all while you're at it

nmattia commented on bin/format-sh:
```diff
@@ -0,0 +1,12 @@
+cd "$(dirname "$(realpath "$0")")/.." || exit
+
+list_files() {
+	git ls-files | grep -E '[.]sh$'
```
so will this file be formatted?

Probably more correct to loop over the files and do something like this:

``` bash
file $file | grep 'Bourne' >/dev/null
```

as we did [in the ic](https://github.com/dfinity-lab/common/blob/05791d3db5c8813e86e194b097624464798b1cf6/pkgs/overlays/lib/default.nix#L114) (private)

bitdivine replied:
Good catch.

Re file, it looks as if that won't be enough, as .sh files won't be checked unless there is a shebang line, and we won't know if there is a missing shebang line on a script unless there is a shebang line on it.
```
$ echo echo echo > whatever.sh
$ file whatever.sh
whatever.sh: ASCII text
```
In my initial filter I also checked the shebang line but I stripped it out, trying to get to something as simple as possible.  It seems that I stripped out too much.  We could check both all files that end in .sh and all that have bash magic strings.

If the shebang is present it works fine, as long as we don't use bash's better cousin. ;-)
```
max@sinkpad:~/dfn/nns-dapp/branches/shellcheck (2:31)$ echo '#!/usr/bin/env bash' > whatever.sh 
max@sinkpad:~/dfn/nns-dapp/branches/shellcheck (3:09)$ file whatever.sh
whatever.sh: Bourne-Again shell script, ASCII text executable
max@sinkpad:~/dfn/nns-dapp/branches/shellcheck (3:09)$ echo '#!/usr/bin/env ksh' > whatever.sh 
max@sinkpad:~/dfn/nns-dapp/branches/shellcheck (5:48)$ file whatever.sh
whatever.sh: a /usr/bin/env ksh script, ASCII text executable
```

So maybe:
```
list_files() {
  git ls-files | while read line ; do if [[ "$line" = *.sh ]] || file "$line" | grep -qw Bourne ; then echo "$line" ; fi ; done
}
```

nmattia replied:
I think it's safe to assume all files have a shebang, and I think it's sensible to limit ourselves to bash. WDYT?

bitdivine replied:
Before this PR these files had no shebang:  `deploy.sh` `... nns-js/proto/update_proto.sh` `frontend/ts/src/proto/update_proto.sh` `frontend/ts/update_candid.sh`  `update_config.sh`

But I do think that sticking to bash is appropriate.

nmattia commented on bin/format-sh:
```diff
@@ -0,0 +1,12 @@
+
+case "${1:-}" in
+--list) list_files ;;
+--check) list_files | xargs -I{} shfmt -i 2 -d "{}" ;;
```
If we want something similar to the II's repo (and what tools like pre-commit do) then we can drop `--check` and always format; then if we want to check we just ask git if anything check. In II we go the extra mile and commit the changed files.

bitdivine replied:
I'm a bit wary but OK with that.  Here all the scripts are run at build time so if formatting causes an error it will probably just show up as a ci eror.

bitdivine replied:
Ok, one more point:  We ave several "format and commit" steps in separate workflows.  If two of them try to commit changes, one will fail.  Better to pull them all out into one workflow.

nmattia commented on Dockerfile:
```diff
@@ -48,7 +48,7 @@ RUN cargo install --version 0.3.1 ic-cdk-optimizer
 
 ENV DEPLOY_ENV=mainnet
-RUN cd frontend/ts && ./build.sh
+RUN cd frontend/ts && bash ./build.sh
```
why call back explicitly here?

nmattia commented on Dockerfile:
```diff
@@ -14,7 +14,7 @@ RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone &
     apt -yqq install --no-install-recommends curl ca-certificates \
         build-essential pkg-config libssl-dev llvm-dev liblmdb-dev clang cmake \
-	git
+	git bash
```
is bash ever not installed on ubuntu?

bitdivine replied:
It's something weird in the docker build that showed up in CI after reformatting ... talk about changes in behaviour after reformatting.  I'll ping you when I have figured out what's going on and you can re-review.

bitdivine replied:
Ah, puzzle solved.  The /bin I added replaces the /bin in the docker ubuntu.  Renaming the local bin to e.g. utils fixes the build issue.  That was really confusing.

nmattia commented on bin/format-sh:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+cd "$(dirname "$(realpath "$0")")/.." || exit
```
I think it's a bit safer to use `BASH_SOURCE` but I might be wrong

bitdivine replied:
`realpath`, or `readlink -f` should resolve symlinks correctly.  See: https://pubs.opengroup.org/onlinepubs/9699919799/functions/realpath.html  Note:  `realpath` has only been added to the linux standard base about 4 years ago.  That said, given that bash is specified in the shebang line above I quite like that solution, and it will work on older systems such as ubuntu 18 which did not have realpath installed by default.  It will fail only if there is code that uses a non-bash interpreter to run a bash script.

bitdivine commented:
@ielashi - Could you re-review please?  The following changes may be of particular interest:
- Combined the github actions formatting steps into a single job, otherwise one of the pushes will fail, if there are several.
- Added cargo fmt to the formatting commands, alongside shfmt and prettier for ts and a special prettier for svelte.
- Removed an empty "generate-wasm.sh" file that was generating linter errors and doesn't seem to be used.

bitdivine commented on .github/workflows/checks.yml:
```diff
@@ -4,10 +4,64 @@ on:
   push:
```
Due to popular demand, `svelte` will be used instead of `svelte-rc1` for all releases.

bitdivine commented on .github/workflows/checks.yml:
```diff
@@ -124,17 +165,13 @@ jobs:
       - name: Run linter
         run: npm run check
 
```
This is new:  Linting with shellcheck.

bitdivine commented on .github/workflows/checks.yml:
```diff
@@ -4,10 +4,64 @@ on:
+
+      - name: Install shfmt
+        run: sudo snap install shfmt
+
```
This is new: The typescript formatting was not being addressed in ci.

bitdivine commented on .github/workflows/checks.yml:
```diff
@@ -4,10 +4,64 @@ on:
+
+      - name: Format rust
+        run: cargo fmt
+
```
Likewise shell scripts are now formatted:

bitdivine commented on Dockerfile:
```diff
@@ -4,6 +4,7 @@
 # docker run --rm --entrypoint cat nns-dapp /nns-dapp.wasm > nns-dapp.wasm
```
This makes sure that bash is used in docker, as everywhere else.

bitdivine commented on build.sh:
```diff
@@ -4,7 +4,7 @@ set -e
 
```
Formatting only, thanks to ci...

nmattia commented on .github/workflows/checks.yml:
```diff
@@ -7,7 +7,60 @@ on:
   pull_request:
 
```
If you wanted, you could have added a new job for committing, and made it depend on the other jobs. Just saying, not a recommendation!

bitdivine replied:
I thought about it.  The only other job is the docker build, which is the slowest job of them all.  I don't want to slow it down further.

bitdivine replied:
Ah, I see what you mean, reversing the order.  I thought about that as well but decided against it.  Formatting is relatively cheap, if there is a syntax error it will catch it early before running more expensive tests.

nmattia commented on .github/workflows/checks.yml:
```diff
@@ -4,10 +4,64 @@ on:
+        run: cargo fmt
+
+      - name: Format shell scripts
+        run: ./util/format-sh
```
nitpick: the scripts directories are usually called `./scripts`, as per GitHub best practices (and the formatting scripts be called `fmt`)

bitdivine replied:
DONE.  I have not renamed the npm formatting commands that both use "format" as that is standard for npm.

nmattia commented on .github/workflows/checks.yml:
```diff
@@ -4,10 +4,64 @@ on:
+
+
   cargo-tests:
+    needs: formatting
```
I think it's very frustrating to start a job, come back an hour later and see it didn't run because of formatting. What is the reason here?

bitdivine replied:
The normal flow would be:
- push
- ci formats
- other tests run, even if formatting had to take place

ielashi commented on .github/workflows/checks.yml:
```diff
@@ -69,21 +124,6 @@ jobs:
         run: npm run test
         working-directory: ./frontend/ts
       
-      - name: Check formatting
```
This was problematic - glad you're removing it.

ielashi reviewed (commented):
LGTM, thanks @bitdivine. I'll final approval to @nmattia.
